### ISSUE TITLE
perf: optimize trace decoding by making identity fetching non-blocking

### ISSUE THREAD

**Issue by mattsse** [foundry-rs/foundry#3236]
### Component

Forge

### Describe the feature you would like

Ref #3006

Decoding traces for better output can make use of etherscan or sig.eth.samczsun.com

both require API requests, the tracing decoding is currently blocking, which means we have code that looks like:

https://github.com/foundry-rs/foundry/blob/903562382075b4ba24e5f4d40aeee620c53a7dde/evm/src/trace/identifier/etherscan.rs#L86

we could solve this by making everything async or use a background task with channels instead

since there's only 1 identifier for the entire suit, the latter makes sense, this way we still keep the non async test interface but can use async decoding

**Comment by zerosnacks** [comment]
Correct me if I'm wrong but I think this is still the case though the code has been updated since: https://github.com/foundry-rs/foundry/blob/626221f5ef44b4af950a08e09bd714650d9eb77d/crates/evm/traces/src/identifier/etherscan.rs#L134-L150

Leaving it open but low priority as this is only really an issue on an initial run, after that the identifiers are loaded from cache.

**Comment by zexoverz** [comment]
This has aged well in parts. Most of what the issue asks for is already in `crates/evm/traces/src/identifier/external.rs`: fetching is async and concurrent, rate-limit aware with backoff, cached per address including negative results, and bounded by a time budget that disables external identification for the rest of the session once it is spent.

The `block_on` at `external.rs:247` is also less severe than it reads. `foundry_common::block_on` is `tokio::task::block_in_place(|| handle.block_on(..))`, so it hands the worker thread back to the runtime rather than stalling it.

What is left is the second half of the original suggestion. `TraceIdentifier::identify_addresses` is still sync, and identification only starts once execution is done. A background task fed by a channel could fetch identities while tests are still running, which is where the remaining wall-clock is.

Making the trait async instead is a wider change than the label suggests: one trait, three production implementors and two in tests, around thirteen call sites, plus the downstream callers in forge and cast. Worth saying that `first issue` no longer fits.

I am happy to implement the background-task version if that is still the direction you want. Would rather ask than open a large refactor you have moved past.

**Comment by mablr** [comment]
@zexoverz Thanks, this is a good summary of how the identifier has evolved. One complication is that Forge already streams completed suites while the runner continues executing later suites, so external identification can overlap with the remaining test execution today. A background task would still need to wait before rendering unless addresses can be submitted earlier than the current `identify_addresses` call.

We should first identify the earlier scheduling point and benchmark the remaining wall-clock benefit on an uncached fork test. That would tell us whether a channel-based refactor is worthwhile. Do you want to start from here ?

**Comment by zexoverz** [comment]
Yes, happy to start there. I will find the earliest point an address set is known ahead of `identify_addresses` and benchmark an uncached fork test against the current streaming behaviour, then bring the numbers back before proposing any refactor.

**Comment by zexoverz** [comment]
Numbers, uncached mainnet fork, 4 suites, 32 external addresses, `-vvvv`, median of 4 runs on an instrumented build of `f854750`. No Etherscan key here so this is the Sourcify fetcher only.

| | default | `--show-progress` |
|---|---|---|
| total wall | 18.0s | 20.6s |
| external identification | 3.9s | 4.4s |
| of that, not hidden by streaming | 1.3s | 4.4s |

You were right, streaming already hides about two thirds of it, and the 1.3s left is the last suite to finish having nothing to overlap against. Earliest complete address set is per test at `runner.rs:1315` and per suite at the setUp trace `runner.rs:803`, both ahead of the `tx.send` at `multi_runner.rs:351` that streaming keys off. Worth noting the whole prize is capped anyway, `remaining_budget` starts at 5s and decrements across the session.

One thing that did come out of it: `--show-progress` collects every suite before sending any at `multi_runner.rs:305-334`, so it overlaps nothing and pays the full cost.

**Comment by zexoverz** [comment]
@mablr those are the uncached fork numbers you asked for, posted above. Want me to open a draft against the earlier scheduling point, or wait for your read on them first?

**Comment by mablr** [comment]
@zexoverz Thanks for following up (sorry for the answer delay). Yes, please open a PR focused on Forge prefetching, keeping the synchronous identifier interface and progress-bar rendering behavior.

Please include the benchmark fixture (in PR body) and reproducible before/after results for both modes. The numbers show potential headroom; let’s see how much earlier scheduling actually recovers while preserving lookup filtering, deduplication, and the timeout budget.
